src/python/createrepo_c/__init__.py

```diff
@@ -382,6 +386,364 @@ def __init__(self, primary_path, filelists_path, other_path, newpkgcb=None, warn
+            ),
+            "filelists": MetadataInfoHolder(
+                fil_xml_path, FilelistsXmlFile(str(fil_xml_path), compressiontype=compression)
+            ),
+            "other": MetadataInfoHolder(
+                oth_xml_path, OtherXmlFile(str(oth_xml_path), compressiontype=compression)
+            ),
+        }
+        self.additional_metadata_files = {}
+
+        if num_packages is not None:
+            self.set_num_of_pkgs(num_packages)
+
+        self._updaterecords = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        if exc_type:
+            pass
+        else:
+            self.finish()
+
+    @property
+    def path(self):
+        return self._destination_repo_path
+
+    @property
+    def repodata_dir(self):
+        return self.path / "repodata"
+
+    def set_num_of_pkgs(self, num):
+        """Set the number of packages that will be added - this has to be done before adding any packages."""
+        assert not self._has_set_num_pkgs, "The number of packages has already been set"
+        self._has_set_num_pkgs = True
+
+        self.working_metadata_files["primary"].writer.set_num_of_pkgs(num)
+        self.working_metadata_files["filelists"].writer.set_num_of_pkgs(num)
+        self.working_metadata_files["other"].writer.set_num_of_pkgs(num)
+
+    def add_pkg_from_file(self, path, output_dir=None):
+        """Add a package to the repo from a provided path."""
+        assert self._has_set_num_pkgs, "Must set the number of packages before adding packages"
+        assert not self._finished, self._FINISHED_ERR_MSG
+
+        filename = os.path.basename(path)
+        try:
+            relative_path = Path(path).relative_to(self.path)  # raises a ValueError if path is not relative
+        except ValueError:
+            if output_dir:
+                relative_path = os.path.join(output_dir, filename)
+                os.makedirs(self.path / output_dir, exist_ok=True)
+            else:
+                relative_path = filename
+            new_path = shutil.copy2(path, self.path / relative_path)
+
+        pkg = package_from_rpm(
+            path,
+            checksum_type=self._checksum_type,
+            location_href=str(relative_path),
+            location_base=None,
+            changelog_limit=self._changelog_limit
+        )
+
+        self.add_pkg(pkg)
+        return pkg
+
+    def add_pkg(self, pkg):
+        """Add a package to the repo from a pre-created Package object."""
+        assert self._has_set_num_pkgs, "Must set the number of packages before adding packages"
+        assert not self._finished, self._FINISHED_ERR_MSG
+
+        self.working_metadata_files["primary"].writer.add_pkg(pkg)
+        self.working_metadata_files["filelists"].writer.add_pkg(pkg)
+        self.working_metadata_files["other"].writer.add_pkg(pkg)
+
+    def add_repomd_metadata(self, name, path, use_compression=True):
```

Comment:
@kontura I didn't notice this until now (sorry), but `add_repomd_metadata` is a bit redundant.  Should we fix that name?  Or is it fine because we're adding metadata to "repomd"?

Reply:
So to change it to just `add_metadata`?
I think I would like that a little bit more but I am also fine with the current form.

If you want to change it I believe we still can but the sooner the better.